# Data Cleaning Code for 2024 and 2023 Sales Data
* Data Cleaning Code
* Data SumarizationMB


## Merge Files for 2024 Sales Data
* Through matching composition key (C1AccountNp & CXRecords)
* Keep rows with 0 sales for consistency

In [1]:
import pandas as pd

In [2]:
fp2024 = pd.read_excel("Goldmine 2024 Sales.xlsx")
print(fp2024.head())


        C1AccountNo        CXRecords  2024 Apr  2024 Jun  2024 Jul  2024 Aug  \
0  A2061733054*,CNR  91470CV(.{82 W<       0.0       0.0       0.0       0.0   
1  A2061733055&RY+5  91470ER(#7ZS W<       0.0       0.0       0.0       0.0   
2  A2061733055&RY+5  9147BAL$5?9W W<       0.0       0.0       0.0       0.0   
3  A2061733055(MK58  91470DV#294$ W<       0.0       0.0       0.0       0.0   
4  A2061733055)S{G]  91470F3(E^Q/ W<       0.0       0.0       0.0       0.0   

   2024 Nov  2024 Oct  2024 May  2024 Dec  2024 Q1  2024 Q2  
0         0         0       0.0         0      0.0      0.0  
1         0         0       0.0         0      0.0      0.0  
2         0         0       0.0         0      0.0      0.0  
3         0         0       0.0         0      0.0      0.0  
4         0         0       0.0         0      0.0      0.0  


In [3]:
sp2024 = pd.read_excel("Goldmine 2023 Q4 and some 2024 Sales.xlsx")
print(sp2024.head())

        C1AccountNo        CXRecords  2023 Q4 PromoStandards Company Store  \
0  A2061733055&RY+5  91470ER(#7ZS W<      0.0            NaN           NaN   
1  A2061733055)S{G]  91470F3(E^Q/ W<      0.0            NaN           NaN   
2  A2061733056$FX$L  91470G4$R)&< W<      0.0            NaN           NaN   
3  A2061733057#*N-2  91470GQ%![I! W<      0.0            NaN           NaN   
4  A2061733058#^F7+  91470IF(9`JJ W<      0.0            NaN           NaN   

   2024 Sales  2024 Jan  2024 Feb  2024 Mar  2024 Sep  
0           0       0.0       0.0       0.0         0  
1           0       0.0       0.0       0.0         0  
2           0       0.0       0.0       0.0         0  
3       44328       0.0       0.0       0.0         0  
4           0       0.0       0.0       0.0         0  


### Initial Dataset Summaries
* Number of Rows
* Numbr of unique keys (composition key for each dataset)
* Whether there are mismatched keys in two dataset: to determine whether there are missing accounts (keys) for some months
* Examine the missing Values (Examine each column with missing values)
* Determine whether the two datasets could be merged without additional manipulation ot information

In [6]:
# Examine the number of rows and columns for each dataset
print(fp2024.shape) 
print(sp2024.shape) 

(124414, 12)
(124414, 10)


In [7]:
# Examine the number of unique keys (C1AccountNo + CXRecords) in each dataset
# Create composite keys for each dataset by combining 'C1AccountNo' and 'CXRecords'
fp2024['composite_key'] = fp2024['C1AccountNo'].astype(str) + '-' + fp2024['CXRecords'].astype(str)
sp2024['composite_key'] = sp2024['C1AccountNo'].astype(str) + '-' + sp2024['CXRecords'].astype(str)

unique_keys_fp2024 = fp2024['composite_key'].nunique()
unique_keys_sp2024 = sp2024['composite_key'].nunique()

print(f"There are {unique_keys_fp2024} unique composite keys in the Goldmine 2024 Sales Dataset")
print(f"There are {unique_keys_sp2024} unique composite keys in the Goldmine 2023 Q4 and some 2024 Sales Dataset")

There are 124414 unique composite keys in the Goldmine 2024 Sales Dataset
There are 124414 unique composite keys in the Goldmine 2023 Q4 and some 2024 Sales Dataset


In [8]:
# Determine whether there are mismatched keys in each dataset (whether one composition key exist in one dataset but not the other one)

# Convert composite keys into sets for easy comparison
keys_fp2024 = set(fp2024['composite_key'])
keys_sp2024 = set(sp2024['composite_key'])

# Find mismatched keys
keys_in_fp_not_in_sp = keys_fp2024 - keys_sp2024  # Keys in fp2024 but not in sp2024
keys_in_sp_not_in_fp = keys_sp2024 - keys_fp2024  # Keys in sp2024 but not in fp2024

if keys_in_fp_not_in_sp:
    print(f"Keys in Goldmine 2024 Sales Dataset but not in Goldmine 2023 Q4 and some 2024 Sales Dataset:")
    print(keys_in_fp_not_in_sp)
else:
    print("No mismatched keys found in Goldmine 2024 Sales Dataset that are missing in the other dataset.")

if keys_in_sp_not_in_fp:
    print(f"Keys in Goldmine 2023 Q4 and some 2024 Sales Dataset but not in Goldmine 2024 Sales Dataset:")
    print(keys_in_sp_not_in_fp)
else:
    print("No mismatched keys found in Goldmine 2023 Q4 and some 2024 Sales Dataset that are missing in the other dataset.")


No mismatched keys found in Goldmine 2024 Sales Dataset that are missing in the other dataset.
No mismatched keys found in Goldmine 2023 Q4 and some 2024 Sales Dataset that are missing in the other dataset.


In [9]:
# Examine the missing values inside the datasets

# Check for missing values in the fp2024 dataset
missing_values_fp2024 = fp2024.isnull().sum()

# Check for missing values in the sp2024 dataset
missing_values_sp2024 = sp2024.isnull().sum()

# Print the number of missing values in each column for both datasets
print("Missing values in Goldmine 2024 Sales Dataset (fp2024):")
print(missing_values_fp2024)

print("\nMissing values in Goldmine 2023 Q4 and some 2024 Sales Dataset (sp2024):")
print(missing_values_sp2024)


""" print("\nColumns with missing values in Goldmine 2024 Sales Dataset (fp2024):")
print(missing_values_fp2024[missing_values_fp2024 > 0])

print("\nColumns with missing values in Goldmine 2023 Q4 and some 2024 Sales Dataset (sp2024):")
print(missing_values_sp2024[missing_values_sp2024 > 0]) """


Missing values in Goldmine 2024 Sales Dataset (fp2024):
C1AccountNo      0
CXRecords        0
2024 Apr         0
2024 Jun         0
2024 Jul         0
2024 Aug         0
2024 Nov         0
2024 Oct         0
2024 May         0
2024 Dec         0
2024 Q1          0
2024 Q2          0
composite_key    0
dtype: int64

Missing values in Goldmine 2023 Q4 and some 2024 Sales Dataset (sp2024):
C1AccountNo            0
CXRecords              0
2023 Q4                0
PromoStandards    123671
Company Store     124412
2024 Sales             0
2024 Jan               0
2024 Feb               0
2024 Mar               0
2024 Sep               0
composite_key          0
dtype: int64


' print("\nColumns with missing values in Goldmine 2024 Sales Dataset (fp2024):")\nprint(missing_values_fp2024[missing_values_fp2024 > 0])\n\nprint("\nColumns with missing values in Goldmine 2023 Q4 and some 2024 Sales Dataset (sp2024):")\nprint(missing_values_sp2024[missing_values_sp2024 > 0]) '

#### Initial Dataset Summaries
* There are 124414 rows in each dataset
* There are 124414 unique composite keys in the Goldmine 2024 Sales Dataset
* There are 124414 unique composite keys in the Goldmine 2023 Q4 and some 2024 Sales Dataset
* No mismatched keys found in the datasets
* The only columns with missing vales is "PromoStandards" and "Company Store", which we will not include in the dataset in our current analysis as the percentage of missing vales is greater that 90%.


### Remove columns and Merge Datasets for Year 2024

* Remove the columns described above that has great percentage of missing values
* Merge the two datasets into a dataframe that comtain the monthly sales for year 2024 

In [10]:
#Remove the two columns in sp2024 dataframe
# Remove the columns 'PromoStandards' and 'Company Store'
sp2024 = sp2024.drop(columns=['PromoStandards', 'Company Store'])

# Print the result to check if the columns have been removed
print(sp2024.head())


        C1AccountNo        CXRecords  2023 Q4  2024 Sales  2024 Jan  2024 Feb  \
0  A2061733055&RY+5  91470ER(#7ZS W<      0.0           0       0.0       0.0   
1  A2061733055)S{G]  91470F3(E^Q/ W<      0.0           0       0.0       0.0   
2  A2061733056$FX$L  91470G4$R)&< W<      0.0           0       0.0       0.0   
3  A2061733057#*N-2  91470GQ%![I! W<      0.0       44328       0.0       0.0   
4  A2061733058#^F7+  91470IF(9`JJ W<      0.0           0       0.0       0.0   

   2024 Mar  2024 Sep                     composite_key  
0       0.0         0  A2061733055&RY+5-91470ER(#7ZS W<  
1       0.0         0  A2061733055)S{G]-91470F3(E^Q/ W<  
2       0.0         0  A2061733056$FX$L-91470G4$R)&< W<  
3       0.0         0  A2061733057#*N-2-91470GQ%![I! W<  
4       0.0         0  A2061733058#^F7+-91470IF(9`JJ W<  


In [11]:
print(fp2024.head())

        C1AccountNo        CXRecords  2024 Apr  2024 Jun  2024 Jul  2024 Aug  \
0  A2061733054*,CNR  91470CV(.{82 W<       0.0       0.0       0.0       0.0   
1  A2061733055&RY+5  91470ER(#7ZS W<       0.0       0.0       0.0       0.0   
2  A2061733055&RY+5  9147BAL$5?9W W<       0.0       0.0       0.0       0.0   
3  A2061733055(MK58  91470DV#294$ W<       0.0       0.0       0.0       0.0   
4  A2061733055)S{G]  91470F3(E^Q/ W<       0.0       0.0       0.0       0.0   

   2024 Nov  2024 Oct  2024 May  2024 Dec  2024 Q1  2024 Q2  \
0         0         0       0.0         0      0.0      0.0   
1         0         0       0.0         0      0.0      0.0   
2         0         0       0.0         0      0.0      0.0   
3         0         0       0.0         0      0.0      0.0   
4         0         0       0.0         0      0.0      0.0   

                      composite_key  
0  A2061733054*,CNR-91470CV(.{82 W<  
1  A2061733055&RY+5-91470ER(#7ZS W<  
2  A2061733055&RY+5-9147BA

In [12]:
# Merge two dataframe by composite_key column
sales2024 = pd.merge(sp2024, fp2024, on='composite_key', how='outer')


print(sales2024.head())
print(sales2024.shape)

      C1AccountNo_x      CXRecords_x  2023 Q4  2024 Sales  2024 Jan  2024 Feb  \
0  A2061733055&RY+5  91470ER(#7ZS W<      0.0           0       0.0       0.0   
1  A2061733055)S{G]  91470F3(E^Q/ W<      0.0           0       0.0       0.0   
2  A2061733056$FX$L  91470G4$R)&< W<      0.0           0       0.0       0.0   
3  A2061733057#*N-2  91470GQ%![I! W<      0.0       44328       0.0       0.0   
4  A2061733058#^F7+  91470IF(9`JJ W<      0.0           0       0.0       0.0   

   2024 Mar  2024 Sep                     composite_key     C1AccountNo_y  \
0       0.0         0  A2061733055&RY+5-91470ER(#7ZS W<  A2061733055&RY+5   
1       0.0         0  A2061733055)S{G]-91470F3(E^Q/ W<  A2061733055)S{G]   
2       0.0         0  A2061733056$FX$L-91470G4$R)&< W<  A2061733056$FX$L   
3       0.0         0  A2061733057#*N-2-91470GQ%![I! W<  A2061733057#*N-2   
4       0.0         0  A2061733058#^F7+-91470IF(9`JJ W<  A2061733058#^F7+   

   ... 2024 Apr  2024 Jun  2024 Jul  2024 Aug  202

In [13]:
sales2024clean = sales2024.drop(columns=['C1AccountNo_y', 'CXRecords_y', '2023 Q4'])
print(sales2024clean.head())
print(sales2024clean.shape)

      C1AccountNo_x      CXRecords_x  2024 Sales  2024 Jan  2024 Feb  \
0  A2061733055&RY+5  91470ER(#7ZS W<           0       0.0       0.0   
1  A2061733055)S{G]  91470F3(E^Q/ W<           0       0.0       0.0   
2  A2061733056$FX$L  91470G4$R)&< W<           0       0.0       0.0   
3  A2061733057#*N-2  91470GQ%![I! W<       44328       0.0       0.0   
4  A2061733058#^F7+  91470IF(9`JJ W<           0       0.0       0.0   

   2024 Mar  2024 Sep                     composite_key  2024 Apr  2024 Jun  \
0       0.0         0  A2061733055&RY+5-91470ER(#7ZS W<       0.0       0.0   
1       0.0         0  A2061733055)S{G]-91470F3(E^Q/ W<       0.0       0.0   
2       0.0         0  A2061733056$FX$L-91470G4$R)&< W<       0.0       0.0   
3       0.0         0  A2061733057#*N-2-91470GQ%![I! W<       0.0       0.0   
4       0.0         0  A2061733058#^F7+-91470IF(9`JJ W<       0.0       0.0   

   2024 Jul  2024 Aug  2024 Nov  2024 Oct  2024 May  2024 Dec  2024 Q1  \
0       0.0       

In [14]:
#Rearrange the columns and rename column names

desired_column_order = [
    'C1AccountNo_x', 'CXRecords_x', 'composite_key', 
    '2024 Jan', '2024 Feb', '2024 Mar', '2024 Apr', '2024 May', 
    '2024 Jun', '2024 Jul', '2024 Aug', '2024 Sep', '2024 Oct', 
    '2024 Nov', '2024 Dec', '2024 Q1', '2024 Q2', '2024 Sales'
]

sales2024clean = sales2024clean[desired_column_order]

sales2024clean = sales2024clean.rename(columns={
    'C1AccountNo_x': 'C1AccountNo',
    'CXRecords_x': 'CXRecords'
})

print(sales2024clean.head())
print(sales2024clean.shape)


        C1AccountNo        CXRecords                     composite_key  \
0  A2061733055&RY+5  91470ER(#7ZS W<  A2061733055&RY+5-91470ER(#7ZS W<   
1  A2061733055)S{G]  91470F3(E^Q/ W<  A2061733055)S{G]-91470F3(E^Q/ W<   
2  A2061733056$FX$L  91470G4$R)&< W<  A2061733056$FX$L-91470G4$R)&< W<   
3  A2061733057#*N-2  91470GQ%![I! W<  A2061733057#*N-2-91470GQ%![I! W<   
4  A2061733058#^F7+  91470IF(9`JJ W<  A2061733058#^F7+-91470IF(9`JJ W<   

   2024 Jan  2024 Feb  2024 Mar  2024 Apr  2024 May  2024 Jun  2024 Jul  \
0       0.0       0.0       0.0       0.0       0.0       0.0       0.0   
1       0.0       0.0       0.0       0.0       0.0       0.0       0.0   
2       0.0       0.0       0.0       0.0       0.0       0.0       0.0   
3       0.0       0.0       0.0       0.0       0.0       0.0       0.0   
4       0.0       0.0       0.0       0.0       0.0       0.0       0.0   

   2024 Aug  2024 Sep  2024 Oct  2024 Nov  2024 Dec  2024 Q1  2024 Q2  \
0       0.0         0         0

## Data Cleaning and Data Question for Year 2024

* Calculate the accurate 2024 Q1 and 2024 Q2 values based on monthly sales
* Determine whether the 2024 Sales (column)'s value is accurate
* record the columns with problem for questions

In [16]:
# Determine wheher there are values diffrent than 0 in the Q1 and Q2 column
#If there are values that are differeht than0, the rows are taken out for further determination
# Filter rows where either '2024 Q1' or '2024 Q2' have values different from 0
non_zero_Q1_Q2 = sales2024clean[(sales2024clean['2024 Q1'] != 0) | (sales2024clean['2024 Q2'] != 0)]

print(len(non_zero_Q1_Q2))
print(non_zero_Q1_Q2.head())

6699
         C1AccountNo        CXRecords                     composite_key  \
5   A2061733058)U90^  91470I8%.+ML W<  A2061733058)U90^-91470I8%.+ML W<   
6   A2061733058)U90^  KIS9VFG)./*97_*  A2061733058)U90^-KIS9VFG)./*97_*   
16  A2061733063(W4CH  91470PA$4!2O W<  A2061733063(W4CH-91470PA$4!2O W<   
37  A2061733076%D/F+  9147174(RL-R W<  A2061733076%D/F+-9147174(RL-R W<   
48  A2061733084%\]$8  91471IO&D/B% W<  A2061733084%\]$8-91471IO&D/B% W<   

    2024 Jan  2024 Feb  2024 Mar  2024 Apr  2024 May  2024 Jun  2024 Jul  \
5     168.25   1677.63      0.00    416.92    472.24    847.03    180.19   
6     168.25   1677.63      0.00    416.92    472.24    847.03    180.19   
16      0.00      0.00      0.00   1389.85      0.00      0.00      0.00   
37      0.00      0.00    599.97      0.00      0.00      0.00      0.00   
48      0.00      0.00   1512.00    201.03      0.00   1134.73     80.37   

    2024 Aug  2024 Sep  2024 Oct  2024 Nov  2024 Dec  2024 Q1  2024 Q2  \
5        0.0 

In [20]:
#Determine whether the values in Q1 and Q2 adds up based on monthly sales
import numpy as np

# Calculate the sum of January, February, and March, and compare it to Q1 using np.isclose
non_zero_Q1_Q2['Q1_check'] = non_zero_Q1_Q2['2024 Jan'] + non_zero_Q1_Q2['2024 Feb'] + non_zero_Q1_Q2['2024 Mar']
non_zero_Q1_Q2['Q1_matches'] = np.isclose(non_zero_Q1_Q2['Q1_check'], non_zero_Q1_Q2['2024 Q1'], atol=1e-2)

# Calculate the sum of April, May, and June, and compare it to Q2 using np.isclose
non_zero_Q1_Q2['Q2_check'] = non_zero_Q1_Q2['2024 Apr'] + non_zero_Q1_Q2['2024 May'] + non_zero_Q1_Q2['2024 Jun']
non_zero_Q1_Q2['Q2_matches'] = np.isclose(non_zero_Q1_Q2['Q2_check'], non_zero_Q1_Q2['2024 Q2'], atol=1e-2)


mismatches = non_zero_Q1_Q2[(non_zero_Q1_Q2['Q1_matches'] == False) | (non_zero_Q1_Q2['Q2_matches'] == False)]

print("Number of rows where Q1 or Q2 values do not match the sum of the respective months:")
print(len(mismatches))


Number of rows where Q1 or Q2 values do not match the sum of the respective months:
3


C:\Users\Michael\AppData\Local\Temp\ipykernel_68184\388178124.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  non_zero_Q1_Q2['Q1_check'] = non_zero_Q1_Q2['2024 Jan'] + non_zero_Q1_Q2['2024 Feb'] + non_zero_Q1_Q2['2024 Mar']
C:\Users\Michael\AppData\Local\Temp\ipykernel_68184\388178124.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  non_zero_Q1_Q2['Q1_matches'] = np.isclose(non_zero_Q1_Q2['Q1_check'], non_zero_Q1_Q2['2024 Q1'], atol=1e-2)
C:\Users\Michael\AppData\Local\Temp\ipykernel_68184\388178124.

In [21]:
print(mismatches.head())

                 C1AccountNo        CXRecords  \
123916  C4013082970%[>JQTLui  OPCW7GN(PR&@>NG   
123917  C4013084154%[>JQTAle  OPCXH4P%&=*=>NG   
123918  C4013084249%[>JQTJen  OPCXKTQ*?J_X>NG   

                               composite_key  2024 Jan  2024 Feb  2024 Mar  \
123916  C4013082970%[>JQTLui-OPCW7GN(PR&@>NG  10075.82       0.0       0.0   
123917  C4013084154%[>JQTAle-OPCXH4P%&=*=>NG  10075.82       0.0       0.0   
123918  C4013084249%[>JQTJen-OPCXKTQ*?J_X>NG  10075.82       0.0       0.0   

        2024 Apr  2024 May  2024 Jun  2024 Jul  ...  2024 Oct  2024 Nov  \
123916       0.0       0.0       0.0       0.0  ...         0         0   
123917       0.0       0.0       0.0       0.0  ...         0         0   
123918       0.0       0.0       0.0       0.0  ...         0         0   

        2024 Dec   2024 Q1  2024 Q2  2024 Sales  Q1_check  Q1_matches  \
123916         0  10005.62      0.0       34593  10075.82       False   
123917         0  10005.62      0.0       3

In [22]:
# Determine the vales in 2024 Sales column adds up based on 2024 monthly sales

# Calculate the sum of all monthly sales (Jan to Dec) and compare it to the '2024 Sales' column
non_zero_Q1_Q2['monthly_sales_sum'] = (
    non_zero_Q1_Q2['2024 Jan'] + non_zero_Q1_Q2['2024 Feb'] + non_zero_Q1_Q2['2024 Mar'] +
    non_zero_Q1_Q2['2024 Apr'] + non_zero_Q1_Q2['2024 May'] + non_zero_Q1_Q2['2024 Jun'] +
    non_zero_Q1_Q2['2024 Jul'] + non_zero_Q1_Q2['2024 Aug'] + non_zero_Q1_Q2['2024 Sep'] +
    non_zero_Q1_Q2['2024 Oct'] + non_zero_Q1_Q2['2024 Nov'] + non_zero_Q1_Q2['2024 Dec']
)

# Compare the sum with the '2024 Sales' column using np.isclose
non_zero_Q1_Q2['sales_matches'] = np.isclose(non_zero_Q1_Q2['monthly_sales_sum'], non_zero_Q1_Q2['2024 Sales'], atol=1e-2)


sales_mismatches = non_zero_Q1_Q2[non_zero_Q1_Q2['sales_matches'] == False]


print("Number of rows where 2024 Sales does not match the sum of monthly sales:")
print(len(sales_mismatches))

Number of rows where 2024 Sales does not match the sum of monthly sales:
6248
                 C1AccountNo        CXRecords  \
5           A2061733058)U90^  91470I8%.+ML W<   
6           A2061733058)U90^  KIS9VFG)./*97_*   
16          A2061733063(W4CH  91470PA$4!2O W<   
37          A2061733076%D/F+  9147174(RL-R W<   
48          A2061733084%\]$8  91471IO&D/B% W<   
...                      ...              ...   
124388  C4072658234%<2\WCJoe  P21Z6GO);JU8^!P   
124396  C4080150053)I5_=0Bra  P2H5YV6*`U)82I>   
124403  C4080734900&[=Y0,Mel  P2W5A9A$*%BMCB\   
124411  C4081262706!_OS:QSus  P39U1OP#_?B2Q(?   
124412  C4081332434%>9ECTAmy  P3BI74J*K[-)2I>   

                               composite_key  2024 Jan  2024 Feb  2024 Mar  \
5           A2061733058)U90^-91470I8%.+ML W<    168.25   1677.63      0.00   
6           A2061733058)U90^-KIS9VFG)./*97_*    168.25   1677.63      0.00   
16          A2061733063(W4CH-91470PA$4!2O W<      0.00      0.00      0.00   
37          A20617330

C:\Users\Michael\AppData\Local\Temp\ipykernel_68184\2352685137.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  non_zero_Q1_Q2['monthly_sales_sum'] = (
C:\Users\Michael\AppData\Local\Temp\ipykernel_68184\2352685137.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  non_zero_Q1_Q2['sales_matches'] = np.isclose(non_zero_Q1_Q2['monthly_sales_sum'], non_zero_Q1_Q2['2024 Sales'], atol=1e-2)


In [29]:
print(sales_mismatches.iloc[2])

C1AccountNo                          A2061733063(W4CH
CXRecords                             91470PA$4!2O W<
composite_key        A2061733063(W4CH-91470PA$4!2O W<
2024 Jan                                          0.0
2024 Feb                                          0.0
2024 Mar                                          0.0
2024 Apr                                      1389.85
2024 May                                          0.0
2024 Jun                                          0.0
2024 Jul                                          0.0
2024 Aug                                          0.0
2024 Sep                                            0
2024 Oct                                            0
2024 Nov                                            0
2024 Dec                                            0
2024 Q1                                           0.0
2024 Q2                                       1389.85
2024 Sales                                       1390
Q1_check                    

In [36]:
# redo the value check because the 2024 Sales is recorded in integer not float

# Convert the '2024 Sales' column back to integer
non_zero_Q1_Q2['2024 Sales'] = non_zero_Q1_Q2['2024 Sales'].astype(int)

non_zero_Q1_Q2['monthly_sales_sum'] = (
    non_zero_Q1_Q2['2024 Jan'] + non_zero_Q1_Q2['2024 Feb'] + non_zero_Q1_Q2['2024 Mar'] +
    non_zero_Q1_Q2['2024 Apr'] + non_zero_Q1_Q2['2024 May'] + non_zero_Q1_Q2['2024 Jun'] +
    non_zero_Q1_Q2['2024 Jul'] + non_zero_Q1_Q2['2024 Aug'] + non_zero_Q1_Q2['2024 Sep'] +
    non_zero_Q1_Q2['2024 Oct'] + non_zero_Q1_Q2['2024 Nov'] + non_zero_Q1_Q2['2024 Dec']
).round().astype(int)  # Round the sum and then convert to integer

# Compare the integer sum with the '2024 Sales' column
non_zero_Q1_Q2['sales_matches'] = non_zero_Q1_Q2['monthly_sales_sum'] == non_zero_Q1_Q2['2024 Sales']

# Print rows where the '2024 Sales' does not match the sum of the monthly sales
sales_mismatches = non_zero_Q1_Q2[non_zero_Q1_Q2['sales_matches'] == False]

# Print the result
print("Number of rows where 2024 Sales does not match the sum of monthly sales (after converting sum to integers):")
print(len(sales_mismatches))
print(sales_mismatches)



Number of rows where 2024 Sales does not match the sum of monthly sales (after converting sum to integers):
3259
                 C1AccountNo        CXRecords  \
118         A2061733082#E72Z  91471G5#H0,D W<   
215         A2061733098*%WL^  A7PE876*S%>HTK@   
269         A2061733094(FC P  91471WP&*GZ& W<   
414         A2061733124(-WN:  914732A(.)WO W<   
471         A2061733130$A#LR  91473AQ%EL`P W<   
...                      ...              ...   
124334  C4062141084!{D5<;Ama  OZJG26M$*(Q>EU7   
124354  C4070156654#^@9TZSte  P09MO71(IRP_CB\   
124368  C4071149064!R&F/TKim  P0Z4GP7$)].,CB\   
124374  C4071840451)Z]2&TGar  P1GVDR5#@\O1@N?   
124386  C4072654029)+QS*&Isa  P21UO9J)-(#C2I>   

                               composite_key  2024 Jan  2024 Feb  2024 Mar  \
118         A2061733082#E72Z-91471G5#H0,D W<      0.00   6970.35   2051.40   
215         A2061733098*%WL^-A7PE876*S%>HTK@   2043.88      0.00    391.20   
269         A2061733094(FC P-91471WP&*GZ& W<    269.71    251.13

C:\Users\Michael\AppData\Local\Temp\ipykernel_68184\3397149230.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  non_zero_Q1_Q2['2024 Sales'] = non_zero_Q1_Q2['2024 Sales'].astype(int)
C:\Users\Michael\AppData\Local\Temp\ipykernel_68184\3397149230.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  non_zero_Q1_Q2['monthly_sales_sum'] = (
C:\Users\Michael\AppData\Local\Temp\ipykernel_68184\3397149230.py:14: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try usin

In [38]:
sales_mismatches.iloc[0]

C1AccountNo                          A2061733082#E72Z
CXRecords                             91471G5#H0,D W<
composite_key        A2061733082#E72Z-91471G5#H0,D W<
2024 Jan                                          0.0
2024 Feb                                      6970.35
2024 Mar                                       2051.4
2024 Apr                                      4253.35
2024 May                                          0.0
2024 Jun                                        143.3
2024 Jul                                      8749.46
2024 Aug                                          0.0
2024 Sep                                            0
2024 Oct                                            0
2024 Nov                                            0
2024 Dec                                            0
2024 Q1                                       9021.75
2024 Q2                                       4396.65
2024 Sales                                      40761
Q1_check                    

In [39]:
print(sales2024clean.shape)
print(mismatches.shape)
print(sales_mismatches.shape)

(124414, 18)
(3, 22)
(3259, 24)


### 2024 Dataset Questions

* There are three rows where either the Q1 or Q2 sum does not match the monthly sum
- Example row: 123916  C4013082970%[>JQTLui  OPCW7GN(PR&@>NG
* There are 3259 rows where the 2024 Sales does not match the monthly sum
- Example row: C1AccountNo                          A2061733082#E72Z
CXRecords                             91471G5#H0,D W<


## Merge Files for 2023 Sales Data
* Through matching composition key (C1AccountNp & CXRecords)
* Keep rows with 0 sales for consistency

In [40]:
fp2023 = pd.read_excel("Goldmine 2023 Sales and Dealer Designation.xlsx")
print(fp2023.head())

        C1AccountNo        CXRecords Trophy Retailer 3  Trophy Retailer 4  \
0  A2061733054*,CNR  91470CV(.{82 W<        1989-01-01                NaN   
1  A2061733055$SC)`  91470EM$BC>5 W<        1989-01-01                NaN   
2  A2061733055&RY+5  91470ER(#7ZS W<        1989-01-01                NaN   
3  A2061733055&RY+5  9147BAL$5?9W W<        1989-01-01                NaN   
4  A2061733055&RY+5  9147BKI(/*KF W<        1989-01-01                NaN   

   Trophy Retailer 5  Trophy Retailer 6  Trophy Retailer 7  Trophy Retailer 8  \
0                NaN                NaN                NaN                NaN   
1                NaN                NaN                NaN                NaN   
2                NaN                NaN                NaN                NaN   
3                NaN                NaN                NaN                NaN   
4                NaN                NaN                NaN                NaN   

   Trophy Retailer 9 Promo Distributor 1  ... 2023

In [41]:
sp2023 = pd.read_excel("Goldmine 2023 Q4 and some 2024 Sales.xlsx")
print(sp2023.head())

        C1AccountNo        CXRecords  2023 Q4 PromoStandards Company Store  \
0  A2061733055&RY+5  91470ER(#7ZS W<      0.0            NaN           NaN   
1  A2061733055)S{G]  91470F3(E^Q/ W<      0.0            NaN           NaN   
2  A2061733056$FX$L  91470G4$R)&< W<      0.0            NaN           NaN   
3  A2061733057#*N-2  91470GQ%![I! W<      0.0            NaN           NaN   
4  A2061733058#^F7+  91470IF(9`JJ W<      0.0            NaN           NaN   

   2024 Sales  2024 Jan  2024 Feb  2024 Mar  2024 Sep  
0           0       0.0       0.0       0.0         0  
1           0       0.0       0.0       0.0         0  
2           0       0.0       0.0       0.0         0  
3       44328       0.0       0.0       0.0         0  
4           0       0.0       0.0       0.0         0  


### Initial Dataset Summaries
* Number of Rows
* Numbr of unique keys (composition key for each dataset)
* Whether there are mismatched keys in two dataset: to determine whether there are missing accounts (keys) for some months
* Examine the missing Values (Examine each column with missing values)
* Determine whether the two datasets could be merged without additional manipulation ot information

In [42]:
# Examine the number of rows and columns for each dataset
print(fp2023.shape) 
print(sp2023.shape) 

# Examine the number of unique keys (C1AccountNo + CXRecords) in each dataset
# Create composite keys for each dataset by combining 'C1AccountNo' and 'CXRecords'
fp2023['composite_key'] = fp2023['C1AccountNo'].astype(str) + '-' + fp2023['CXRecords'].astype(str)
sp2023['composite_key'] = sp2023['C1AccountNo'].astype(str) + '-' + sp2023['CXRecords'].astype(str)

unique_keys_fp2023 = fp2023['composite_key'].nunique()
unique_keys_sp2023 = sp2023['composite_key'].nunique()

print(f"There are {unique_keys_fp2023} unique composite keys in the Goldmine 2023 Sales and Dealer Designation Dataset")
print(f"There are {unique_keys_sp2023} unique composite keys in the Goldmine 2023 Q4 and some 2024 Sales Dataset")

# Determine whether there are mismatched keys in each dataset (whether one composition key exist in one dataset but not the other one)

# Convert composite keys into sets for easy comparison
keys_fp2023 = set(fp2023['composite_key'])
keys_sp2023 = set(sp2023['composite_key'])

# Find mismatched keys
keys_in_fp_not_in_sp_2023 = keys_fp2023 - keys_sp2023  # Keys in fp2023 but not in sp2023
keys_in_sp_not_in_fp_2023 = keys_sp2023 - keys_fp2023  # Keys in sp2023 but not in fp2023

if keys_in_fp_not_in_sp_2023:
    print(f"Keys in Goldmine 2023 Sales and Dealer Designation Dataset but not in Goldmine 2023 Q4 and some 2024 Sales Dataset:")
    print(keys_in_fp_not_in_sp_2023)
else:
    print("No mismatched keys found in Goldmine 2023 Sales and Dealer Designation Dataset that are missing in the other dataset.")

if keys_in_sp_not_in_fp_2023:
    print(f"Keys in Goldmine 2023 Q4 and some 2024 Sales Dataset but not in Goldmine 2023 Sales and Dealer Designation Dataset:")
    print(keys_in_sp_not_in_fp_2023)
else:
    print("No mismatched keys found in Goldmine 2023 Q4 and some 2024 Sales Dataset that are missing in the other dataset.")

(124414, 32)
(124414, 10)
There are 124414 unique composite keys in the Goldmine 2023 Sales and Dealer Designation Dataset
There are 124414 unique composite keys in the Goldmine 2023 Q4 and some 2024 Sales Dataset
No mismatched keys found in Goldmine 2023 Sales and Dealer Designation Dataset that are missing in the other dataset.
No mismatched keys found in Goldmine 2023 Q4 and some 2024 Sales Dataset that are missing in the other dataset.


In [43]:
# Examine the missing values inside the datasets

# Check for missing values in the fp2024 dataset
missing_values_fp2023 = fp2023.isnull().sum()

# Check for missing values in the sp2024 dataset
missing_values_sp2023 = sp2023.isnull().sum()

# Print the number of missing values in each column for both datasets
print("Missing values in Goldmine 2023 Sales and Dealer Designation Dataset (fp2023):")
print(missing_values_fp2023)

print("\nMissing values in Goldmine 2023 Q4 and some 2024 Sales Dataset (sp2023):")
print(missing_values_sp2023)


""" print("\nColumns with missing values in Goldmine 2023 Sales and Dealer Designation Dataset (fp2023):")
print(missing_values_fp2024[missing_values_fp2023 > 0])

print("\nColumns with missing values in Goldmine 2023 Q4 and some 2024 Sales Dataset (sp2023):")
print(missing_values_sp2024[missing_values_sp2023 > 0]) """


Missing values in Goldmine 2023 Sales and Dealer Designation Dataset (fp2023):
C1AccountNo                       0
CXRecords                         0
Trophy Retailer 3                 0
Trophy Retailer 4            124414
Trophy Retailer 5            124414
Trophy Retailer 6            124414
Trophy Retailer 7            124414
Trophy Retailer 8            124414
Trophy Retailer 9            124414
Promo Distributor 1          124413
Promo Distributor 2          124413
Promo Distributor 3          124413
Promo Distributor 4          124413
Dealer Designation           123586
Individual Strategic Acct    122363
2023 Individual                   0
2023 Sales                        0
2023 Jan                          0
2023 Feb                          0
2023 Mar                          0
2023 Apr                          0
2023 May                          0
2023 Jun                          0
2023 Jul                          0
2023 Aug                          0
2023 Sep             

' print("\nColumns with missing values in Goldmine 2023 Sales and Dealer Designation Dataset (fp2023):")\nprint(missing_values_fp2024[missing_values_fp2023 > 0])\n\nprint("\nColumns with missing values in Goldmine 2023 Q4 and some 2024 Sales Dataset (sp2023):")\nprint(missing_values_sp2024[missing_values_sp2023 > 0]) '

In [44]:
print("\nMissing values in Goldmine 2023 Q4 and some 2024 Sales Dataset (sp2023):")
print(missing_values_sp2023)


Missing values in Goldmine 2023 Q4 and some 2024 Sales Dataset (sp2023):
C1AccountNo            0
CXRecords              0
2023 Q4                0
PromoStandards    123671
Company Store     124412
2024 Sales             0
2024 Jan               0
2024 Feb               0
2024 Mar               0
2024 Sep               0
composite_key          0
dtype: int64


In [45]:
print(fp2023["2023 Individual"])

0         0.0
1         0.0
2         0.0
3         0.0
4         0.0
         ... 
124409    0.0
124410    0.0
124411    0.0
124412    0.0
124413    0.0
Name: 2023 Individual, Length: 124414, dtype: float64


#### Initial Dataset Summaries
* There are 124414 rows in each dataset
* There are 124414 unique composite keys in the Goldmine 2023 Sales and Dealer Designation Dataset
* There are 124414 unique composite keys in the Goldmine 2023 Q4 and some 2024 Sales Dataset
* No mismatched keys found in the datasets
* There are a few columns that needs to be removed since the number of missing values for these columns are greater than 90%

### Remove columns and Merge Datasets for Year 2023

* Remove the columns described above that has great percentage of missing values
* Merge the two datasets into a dataframe that comtain the monthly sales for year 2023
Dealer Designation           123586
Individual Strategic Acct    122363
2023 Individual                   0

In [46]:
#Remove the columns in fp2023 dataframe
fp2023 = fp2023.drop(columns=['Trophy Retailer 3', 'Trophy Retailer 4', 'Trophy Retailer 5', 'Trophy Retailer 6', 'Trophy Retailer 7', 'Trophy Retailer 8', 'Trophy Retailer 9', 'Promo Distributor 1', 'Promo Distributor 2', 'Promo Distributor 3', 'Promo Distributor 4', 'Dealer Designation', 'Individual Strategic Acct', '2023 Individual'])

# Print the result to check if the columns have been removed
print(fp2023.head())

        C1AccountNo        CXRecords  2023 Sales  2023 Jan  2023 Feb  \
0  A2061733054*,CNR  91470CV(.{82 W<           0       0.0       0.0   
1  A2061733055$SC)`  91470EM$BC>5 W<           0       0.0       0.0   
2  A2061733055&RY+5  91470ER(#7ZS W<           0       0.0       0.0   
3  A2061733055&RY+5  9147BAL$5?9W W<           0       0.0       0.0   
4  A2061733055&RY+5  9147BKI(/*KF W<           0       0.0       0.0   

   2023 Mar  2023 Apr  2023 May  2023 Jun  2023 Jul  2023 Aug  2023 Sep  \
0       0.0       0.0       0.0       0.0       0.0       0.0       0.0   
1       0.0       0.0       0.0       0.0       0.0       0.0       0.0   
2       0.0       0.0       0.0       0.0       0.0       0.0       0.0   
3       0.0       0.0       0.0       0.0       0.0       0.0       0.0   
4       0.0       0.0       0.0       0.0       0.0       0.0       0.0   

   2023 Oct  2023 Nov  2023 Dec  2023 Q1  2023 Q2  2023 Q3  \
0       0.0       0.0       0.0      0.0      0.0     

In [47]:
# Keep only the specified columns
sp2023 = sp2023[['C1AccountNo', 'CXRecords', '2023 Q4']]

# Print the result to verify the DataFrame now contains only these columns
print(sp2023.head())

        C1AccountNo        CXRecords  2023 Q4
0  A2061733055&RY+5  91470ER(#7ZS W<      0.0
1  A2061733055)S{G]  91470F3(E^Q/ W<      0.0
2  A2061733056$FX$L  91470G4$R)&< W<      0.0
3  A2061733057#*N-2  91470GQ%![I! W<      0.0
4  A2061733058#^F7+  91470IF(9`JJ W<      0.0


In [48]:
sp2023['composite_key'] = sp2023['C1AccountNo'].astype(str) + '-' + sp2023['CXRecords'].astype(str)
print(sp2023.head())

        C1AccountNo        CXRecords  2023 Q4  \
0  A2061733055&RY+5  91470ER(#7ZS W<      0.0   
1  A2061733055)S{G]  91470F3(E^Q/ W<      0.0   
2  A2061733056$FX$L  91470G4$R)&< W<      0.0   
3  A2061733057#*N-2  91470GQ%![I! W<      0.0   
4  A2061733058#^F7+  91470IF(9`JJ W<      0.0   

                      composite_key  
0  A2061733055&RY+5-91470ER(#7ZS W<  
1  A2061733055)S{G]-91470F3(E^Q/ W<  
2  A2061733056$FX$L-91470G4$R)&< W<  
3  A2061733057#*N-2-91470GQ%![I! W<  
4  A2061733058#^F7+-91470IF(9`JJ W<  


In [49]:
# Merge two dataframe by composite_key column
sales2023 = pd.merge(sp2023, fp2023, on='composite_key', how='outer')


print(sales2023.head())
print(sales2023.shape)

      C1AccountNo_x      CXRecords_x  2023 Q4  \
0  A2061733055&RY+5  91470ER(#7ZS W<      0.0   
1  A2061733055)S{G]  91470F3(E^Q/ W<      0.0   
2  A2061733056$FX$L  91470G4$R)&< W<      0.0   
3  A2061733057#*N-2  91470GQ%![I! W<      0.0   
4  A2061733058#^F7+  91470IF(9`JJ W<      0.0   

                      composite_key     C1AccountNo_y      CXRecords_y  \
0  A2061733055&RY+5-91470ER(#7ZS W<  A2061733055&RY+5  91470ER(#7ZS W<   
1  A2061733055)S{G]-91470F3(E^Q/ W<  A2061733055)S{G]  91470F3(E^Q/ W<   
2  A2061733056$FX$L-91470G4$R)&< W<  A2061733056$FX$L  91470G4$R)&< W<   
3  A2061733057#*N-2-91470GQ%![I! W<  A2061733057#*N-2  91470GQ%![I! W<   
4  A2061733058#^F7+-91470IF(9`JJ W<  A2061733058#^F7+  91470IF(9`JJ W<   

   2023 Sales  2023 Jan  2023 Feb  2023 Mar  ...  2023 Jun  2023 Jul  \
0           0      0.00      0.00      0.00  ...       0.0       0.0   
1           0      0.00      0.00      0.00  ...       0.0       0.0   
2           0      0.00      0.00      0.00 

In [50]:
# Remove the columns 'C1AccountNo_y' and 'CXRecords_y'
sales2023 = sales2023.drop(columns=['C1AccountNo_y', 'CXRecords_y'])

print(sales2023.head())
print(sales2023.shape)

      C1AccountNo_x      CXRecords_x  2023 Q4  \
0  A2061733055&RY+5  91470ER(#7ZS W<      0.0   
1  A2061733055)S{G]  91470F3(E^Q/ W<      0.0   
2  A2061733056$FX$L  91470G4$R)&< W<      0.0   
3  A2061733057#*N-2  91470GQ%![I! W<      0.0   
4  A2061733058#^F7+  91470IF(9`JJ W<      0.0   

                      composite_key  2023 Sales  2023 Jan  2023 Feb  2023 Mar  \
0  A2061733055&RY+5-91470ER(#7ZS W<           0      0.00      0.00      0.00   
1  A2061733055)S{G]-91470F3(E^Q/ W<           0      0.00      0.00      0.00   
2  A2061733056$FX$L-91470G4$R)&< W<           0      0.00      0.00      0.00   
3  A2061733057#*N-2-91470GQ%![I! W<       55561  14960.72   1582.26   2425.43   
4  A2061733058#^F7+-91470IF(9`JJ W<           0      0.00      0.00      0.00   

   2023 Apr  2023 May  2023 Jun  2023 Jul  2023 Aug  2023 Sep  2023 Oct  \
0      0.00      0.00       0.0       0.0       0.0       0.0       0.0   
1      0.00      0.00       0.0       0.0       0.0       0.0       

In [51]:
# Define the desired order of columns
desired_columns_2023 = [
    'C1AccountNo_x', 'CXRecords_x', 'composite_key', '2023 Jan', '2023 Feb', '2023 Mar',
    '2023 Apr', '2023 May', '2023 Jun', '2023 Jul', '2023 Aug', '2023 Sep', '2023 Oct', 
    '2023 Nov', '2023 Dec', '2023 Q1', '2023 Q2', '2023 Q3', '2023 Q4', '2023 Sales'
]

# Reorder the DataFrame columns
sales2023 = sales2023[desired_columns_2023]

print(sales2023.head())
print(sales2023.shape)


      C1AccountNo_x      CXRecords_x                     composite_key  \
0  A2061733055&RY+5  91470ER(#7ZS W<  A2061733055&RY+5-91470ER(#7ZS W<   
1  A2061733055)S{G]  91470F3(E^Q/ W<  A2061733055)S{G]-91470F3(E^Q/ W<   
2  A2061733056$FX$L  91470G4$R)&< W<  A2061733056$FX$L-91470G4$R)&< W<   
3  A2061733057#*N-2  91470GQ%![I! W<  A2061733057#*N-2-91470GQ%![I! W<   
4  A2061733058#^F7+  91470IF(9`JJ W<  A2061733058#^F7+-91470IF(9`JJ W<   

   2023 Jan  2023 Feb  2023 Mar  2023 Apr  2023 May  2023 Jun  2023 Jul  \
0      0.00      0.00      0.00      0.00      0.00       0.0       0.0   
1      0.00      0.00      0.00      0.00      0.00       0.0       0.0   
2      0.00      0.00      0.00      0.00      0.00       0.0       0.0   
3  14960.72   1582.26   2425.43   4863.82  13153.96       0.0       0.0   
4      0.00      0.00      0.00      0.00      0.00       0.0       0.0   

   2023 Aug  2023 Sep  2023 Oct  2023 Nov  2023 Dec   2023 Q1   2023 Q2  \
0       0.0       0.0       0

In [52]:
#rename the first two columns 
sales2023 = sales2023.rename(columns={
    'C1AccountNo_x': 'C1AccountNo',
    'CXRecords_x': 'CXRecords'
})

print(sales2023.head())
print(sales2023.shape)

        C1AccountNo        CXRecords                     composite_key  \
0  A2061733055&RY+5  91470ER(#7ZS W<  A2061733055&RY+5-91470ER(#7ZS W<   
1  A2061733055)S{G]  91470F3(E^Q/ W<  A2061733055)S{G]-91470F3(E^Q/ W<   
2  A2061733056$FX$L  91470G4$R)&< W<  A2061733056$FX$L-91470G4$R)&< W<   
3  A2061733057#*N-2  91470GQ%![I! W<  A2061733057#*N-2-91470GQ%![I! W<   
4  A2061733058#^F7+  91470IF(9`JJ W<  A2061733058#^F7+-91470IF(9`JJ W<   

   2023 Jan  2023 Feb  2023 Mar  2023 Apr  2023 May  2023 Jun  2023 Jul  \
0      0.00      0.00      0.00      0.00      0.00       0.0       0.0   
1      0.00      0.00      0.00      0.00      0.00       0.0       0.0   
2      0.00      0.00      0.00      0.00      0.00       0.0       0.0   
3  14960.72   1582.26   2425.43   4863.82  13153.96       0.0       0.0   
4      0.00      0.00      0.00      0.00      0.00       0.0       0.0   

   2023 Aug  2023 Sep  2023 Oct  2023 Nov  2023 Dec   2023 Q1   2023 Q2  \
0       0.0       0.0       0

## Data Cleaning and Data Question for Year 2023

* Calculate the accurate 2023 quarter sales values based on monthly sales
* Determine whether the 2023 Sales (column)'s value is accurate
* record the columns with problem for questions

In [65]:
# Create a copy of the sales2023 DataFrame
sales2023_copy = sales2023.copy()

# Calculate the sum for each quarter
sales2023_copy['Q1_check'] = sales2023_copy['2023 Jan'] + sales2023_copy['2023 Feb'] + sales2023_copy['2023 Mar']
sales2023_copy['Q2_check'] = sales2023_copy['2023 Apr'] + sales2023_copy['2023 May'] + sales2023_copy['2023 Jun']
sales2023_copy['Q3_check'] = sales2023_copy['2023 Jul'] + sales2023_copy['2023 Aug'] + sales2023_copy['2023 Sep']
sales2023_copy['Q4_check'] = sales2023_copy['2023 Oct'] + sales2023_copy['2023 Nov'] + sales2023_copy['2023 Dec']

# Use np.isclose() to compare the quarterly values with the sums, allowing for small differences
sales2023_copy['Q1_matches'] = np.isclose(sales2023_copy['2023 Q1'], sales2023_copy['Q1_check'], atol=1e-2)
sales2023_copy['Q2_matches'] = np.isclose(sales2023_copy['2023 Q2'], sales2023_copy['Q2_check'], atol=1e-2)
sales2023_copy['Q3_matches'] = np.isclose(sales2023_copy['2023 Q3'], sales2023_copy['Q3_check'], atol=1e-2)
sales2023_copy['Q4_matches'] = np.isclose(sales2023_copy['2023 Q4'], sales2023_copy['Q4_check'], atol=1e-2)

# Filter rows where any of the quarterly values do not match the sum of the corresponding months
mismatched_rows = sales2023_copy[
    (sales2023_copy['Q1_matches'] == False) | 
    (sales2023_copy['Q2_matches'] == False) | 
    (sales2023_copy['Q3_matches'] == False) | 
    (sales2023_copy['Q4_matches'] == False)
]

# Print the number of mismatched rows and the DataFrame for analysis
print("Number of rows with mismatched quarterly values:")
print(len(mismatched_rows))
print(mismatched_rows)

Number of rows with mismatched quarterly values:
14294
                 C1AccountNo        CXRecords  \
117         A2061733081)\<@Q  91471EN)B59N W<   
168         A2061733081)\<@Q  CDVN91I+*H$+3{G   
3220    A2061742370!),0QTMas  914GZSJ%JB\`T&@   
3221    A2061742370$;!6?TMas  914GZS4!?,81T&@   
3223    A2061742370$N91 TMas  9S58DYS(F:\, YU   
...                      ...              ...   
124321  C4061238623(`(A2THan  OYW838A(VLYHEU7   
124337  C4062436718 [9[:!Eli  OZR15PZ$K31`CB\   
124361  C4070835240%.Y>B&Kar  P0QZVDW ONY{^!P   
124363  C4070839854&C)H@TTon  P0R4TDK)N>[DEU7   
124364  C4070859961(&R[67Dia  P0RQD42%,IFMCB\   

                               composite_key   2023 Jan  2023 Feb  2023 Mar  \
117         A2061733081)\<@Q-91471EN)B59N W<       0.00      0.00      0.00   
168         A2061733081)\<@Q-CDVN91I+*H$+3{G       0.00      0.00      0.00   
3220    A2061742370!),0QTMas-914GZSJ%JB\`T&@       0.00      0.00      0.00   
3221    A2061742370$;!6?TMas-914GZS4!?,8

In [66]:
print(mismatched_rows.iloc[0])

C1AccountNo                      A2061733081)\<@Q
CXRecords                         91471EN)B59N W<
composite_key    A2061733081)\<@Q-91471EN)B59N W<
2023 Jan                                      0.0
2023 Feb                                      0.0
2023 Mar                                      0.0
2023 Apr                                      0.0
2023 May                                    227.2
2023 Jun                                      0.0
2023 Jul                                      0.0
2023 Aug                                      0.0
2023 Sep                                      0.0
2023 Oct                                      0.0
2023 Nov                                      0.0
2023 Dec                                      0.0
2023 Q1                                       0.0
2023 Q2                                       0.0
2023 Q3                                       0.0
2023 Q4                                       0.0
2023 Sales                                      0


In [67]:
# Calculate the sum of all 2023 monthly sales (Jan to Dec)
sales2023_copy['monthly_sales_sum'] = (
    sales2023_copy['2023 Jan'] + sales2023_copy['2023 Feb'] + sales2023_copy['2023 Mar'] +
    sales2023_copy['2023 Apr'] + sales2023_copy['2023 May'] + sales2023_copy['2023 Jun'] +
    sales2023_copy['2023 Jul'] + sales2023_copy['2023 Aug'] + sales2023_copy['2023 Sep'] +
    sales2023_copy['2023 Oct'] + sales2023_copy['2023 Nov'] + sales2023_copy['2023 Dec']
)

# Round the sum to the nearest integer (0.5 rounding up)
sales2023_copy['monthly_sales_sum_rounded'] = sales2023_copy['monthly_sales_sum'].round()

# Convert the rounded sum to integer
sales2023_copy['monthly_sales_sum_rounded'] = sales2023_copy['monthly_sales_sum_rounded'].astype(int)

# Compare the rounded sum with the '2023 Sales' column
sales2023_copy['sales_matches'] = sales2023_copy['monthly_sales_sum_rounded'] == sales2023_copy['2023 Sales']

# Filter rows where the '2023 Sales' does not match the sum of the monthly sales
mismatched_rows_2023sales = sales2023_copy[sales2023_copy['sales_matches'] == False]

# Print the number of mismatched rows and the DataFrame for analysis
print("Number of rows where 2023 Sales does not match the sum of monthly sales:")
print(len(mismatched_rows_2023sales))
print(mismatched_rows_2023sales.head())

Number of rows where 2023 Sales does not match the sum of monthly sales:
49592
         C1AccountNo        CXRecords                     composite_key  \
3   A2061733057#*N-2  91470GQ%![I! W<  A2061733057#*N-2-91470GQ%![I! W<   
9   A2061733059+!1ZH  91470KD*/LZZ W<  A2061733059+!1ZH-91470KD*/LZZ W<   
12  A2061733060)(W<%  91470M0#>H/T W<  A2061733060)(W<%-91470M0#>H/T W<   
14  A2061733061&N&YD  91470MX+MQ)I W<  A2061733061&N&YD-91470MX+MQ)I W<   
18  A2061733063)F3!_  91470OV)-![X W<  A2061733063)F3!_-91470OV)-![X W<   

    2023 Jan  2023 Feb  2023 Mar  2023 Apr  2023 May  2023 Jun  2023 Jul  ...  \
3   14960.72   1582.26   2425.43   4863.82  13153.96       0.0       0.0  ...   
9       0.00      0.00      0.00      0.00      0.00       0.0       0.0  ...   
12      0.00   1015.20      0.00      0.00    312.60       0.0       0.0  ...   
14      0.00      0.00      0.00      0.00      0.00       0.0       0.0  ...   
18   1760.12    755.61   2794.39    724.50    602.55       0.0   

In [71]:
print(mismatched_rows_2023sales.iloc[1])

C1AccountNo                                  A2061733059+!1ZH
CXRecords                                     91470KD*/LZZ W<
composite_key                A2061733059+!1ZH-91470KD*/LZZ W<
2023 Jan                                                  0.0
2023 Feb                                                  0.0
2023 Mar                                                  0.0
2023 Apr                                                  0.0
2023 May                                                  0.0
2023 Jun                                                  0.0
2023 Jul                                                  0.0
2023 Aug                                                  0.0
2023 Sep                                                  0.0
2023 Oct                                                  0.0
2023 Nov                                                  0.0
2023 Dec                                                  0.0
2023 Q1                                                   0.0
2023 Q2 

### 2023 Dataset Questions

* There are mismatch rows with 2023 May sales of 227.2, which the record is believed to be incorrect
* There are currently 14294 mismatched rows based on quartly sales comparision
- Example row: C1AccountNo                      A2061742373&AF:)TMas CXRecords  914GZW7(V&3UT&@
- Example row: C1AccountNo                      A2061733081)\<@Q     CXRecords  91471EN)B59N W<
* There are 49592 rows where the 2023 Sales does not match the monthly sum
- Example row: C1AccountNo                      A2061733078$YA`?     CXRecords  914719S%6VWJ W<     
- Example row: C1AccountNo                      A2061733059+!1ZH     CXRecords  91470KD*/LZZ W<

## Examine the output 2023 2024 datasets and export as excel

In [72]:
print(sales2024clean.head())
print(sales2024clean.shape)

        C1AccountNo        CXRecords                     composite_key  \
0  A2061733055&RY+5  91470ER(#7ZS W<  A2061733055&RY+5-91470ER(#7ZS W<   
1  A2061733055)S{G]  91470F3(E^Q/ W<  A2061733055)S{G]-91470F3(E^Q/ W<   
2  A2061733056$FX$L  91470G4$R)&< W<  A2061733056$FX$L-91470G4$R)&< W<   
3  A2061733057#*N-2  91470GQ%![I! W<  A2061733057#*N-2-91470GQ%![I! W<   
4  A2061733058#^F7+  91470IF(9`JJ W<  A2061733058#^F7+-91470IF(9`JJ W<   

   2024 Jan  2024 Feb  2024 Mar  2024 Apr  2024 May  2024 Jun  2024 Jul  \
0       0.0       0.0       0.0       0.0       0.0       0.0       0.0   
1       0.0       0.0       0.0       0.0       0.0       0.0       0.0   
2       0.0       0.0       0.0       0.0       0.0       0.0       0.0   
3       0.0       0.0       0.0       0.0       0.0       0.0       0.0   
4       0.0       0.0       0.0       0.0       0.0       0.0       0.0   

   2024 Aug  2024 Sep  2024 Oct  2024 Nov  2024 Dec  2024 Q1  2024 Q2  \
0       0.0         0         0

In [73]:
print(sales2023.head())
print(sales2023.shape)

        C1AccountNo        CXRecords                     composite_key  \
0  A2061733055&RY+5  91470ER(#7ZS W<  A2061733055&RY+5-91470ER(#7ZS W<   
1  A2061733055)S{G]  91470F3(E^Q/ W<  A2061733055)S{G]-91470F3(E^Q/ W<   
2  A2061733056$FX$L  91470G4$R)&< W<  A2061733056$FX$L-91470G4$R)&< W<   
3  A2061733057#*N-2  91470GQ%![I! W<  A2061733057#*N-2-91470GQ%![I! W<   
4  A2061733058#^F7+  91470IF(9`JJ W<  A2061733058#^F7+-91470IF(9`JJ W<   

   2023 Jan  2023 Feb  2023 Mar  2023 Apr  2023 May  2023 Jun  2023 Jul  \
0      0.00      0.00      0.00      0.00      0.00       0.0       0.0   
1      0.00      0.00      0.00      0.00      0.00       0.0       0.0   
2      0.00      0.00      0.00      0.00      0.00       0.0       0.0   
3  14960.72   1582.26   2425.43   4863.82  13153.96       0.0       0.0   
4      0.00      0.00      0.00      0.00      0.00       0.0       0.0   

   2023 Aug  2023 Sep  2023 Oct  2023 Nov  2023 Dec   2023 Q1   2023 Q2  \
0       0.0       0.0       0

In [76]:
#Export bnoth dataframes to excel files
file_name_2023 = 'Crystal D 2023 sales.xlsx'
file_name_2024 = 'Crystal D 2024 sales.xlsx'
sales2023.to_excel(file_name_2023, index=False)
sales2024clean.to_excel(file_name_2024, index=False)

print('DataFrame is written to Excel File successfully.')

DataFrame is written to Excel File successfully.
